## Diabetes dataset

In [3]:
import pandas as pd
# Load diabetes dataset

diabetes_data = pd.read_csv('diabetes.csv')

## EDA


In [4]:
diabetes_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

### About the dataset
 Here's a breakdown of the columns:

* Pregnancies: Number of times pregnant
* Glucose: Plasma glucose concentration in an oral glucose tolerance test
* BloodPressure: Diastolic blood pressure (mm Hg)
* SkinThickness: Triceps skinfold thickness (mm)
* Insulin: 2-Hour serum insulin (mu U/ml)
* BMI: Body mass index (weight in kg/(height in m)^2)
* DiabetesPedigreeFunction: Diabetes pedigree function (a measure of the diabetes genetic influence)
* Age: Age (years)
* Outcome: Binary variable indicating whether the individual has diabetes (1) or not (0)

In [5]:
# dataset shape
diabetes_data.shape

(768, 9)

In [6]:
# dataset statistical information
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

### Inference
* The features have significantly different scales thus begs for standardization of the dataset.

In [7]:
# dataset features information
diabetes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


### Inferences
* Dataset has 768 rows and 9 columns in total
* All columns are numerical
* There are no null values in all the columns


In [14]:
diabetes_data.Outcome.value_counts()

0    500
1    268
Name: Outcome, dtype: int64

### Inferences
* There are two classes in the target column (Outcome)i.e
* 0 --> Non_diabetic
* 1 --> Diabetic

In [15]:
# Comparison of each feature based on the two classes of the target column.
diabetes_data.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

## Preprocessing

#### Target separation


In [33]:
X = diabetes_data.drop('Outcome',axis = 1)

y = diabetes_data.Outcome


#### Standardization

In [30]:
# Importing the scaler function
from sklearn.preprocessing import StandardScaler

# Initializing the StandardScaler
scaler = StandardScaler()

# Standardizing the features in X
X_standardized = scaler.fit_transform(X)

# Print the standardized X
print(X_standardized)


[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


####Train-test split

In [36]:
# Importing the train_test_split function
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_standardized, y, test_size=0.2, random_state=42, stratify=y)

# Print the shapes of the training and testing sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (614, 8)
X_test shape: (154, 8)
y_train shape: (614,)
y_test shape: (154,)


#### Multi-Model training

In [40]:
# Importing models and evaluation metric
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score


# Initialize and train multiple models
models = {
    'Logistic Regression': LogisticRegression(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'SVM': SVC(),
    'K-Nearest Neighbors': KNeighborsClassifier()
}

for model_name, model in models.items():
    model.fit(X_train, y_train)
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    train_accuracy = accuracy_score(y_train, y_train_pred)
    test_accuracy = accuracy_score(y_test, y_test_pred)
    average_accuracy = (train_accuracy + test_accuracy) / 2
    print(f'{model_name} Train Accuracy: {train_accuracy:.4f}')
    print(f'{model_name} Test Accuracy: {test_accuracy:.4f}')
    print(f'{model_name} Average Accuracy: {average_accuracy:.4f}')
    print()


Logistic Regression Train Accuracy: 0.7915
Logistic Regression Test Accuracy: 0.7143
Logistic Regression Average Accuracy: 0.7529

Decision Tree Train Accuracy: 1.0000
Decision Tree Test Accuracy: 0.7273
Decision Tree Average Accuracy: 0.8636

Random Forest Train Accuracy: 1.0000
Random Forest Test Accuracy: 0.7208
Random Forest Average Accuracy: 0.8604

SVM Train Accuracy: 0.8355
SVM Test Accuracy: 0.7468
SVM Average Accuracy: 0.7911

K-Nearest Neighbors Train Accuracy: 0.8029
K-Nearest Neighbors Test Accuracy: 0.7078
K-Nearest Neighbors Average Accuracy: 0.7554



* The model with the best average accuracy is Decision Tree with Average Accuracy: 0.8636,however, it's Test Accuracy of 0.7273 is beaten by that of SVM.
* SVM Train Accuracy: 0.8355
* SVM Test Accuracy: 0.7468
* SVM Average Accuracy: 0.7911
* Therefore, for this project,i'll use SVM model  for prediction system

#### Predictive system

In [58]:
import numpy as np

# Function to make predictions
def predict_sample(sample):
    # Convert the sample to a 2D array
    sample = np.array([sample])

    # Make predictions for the sample
    SVM = models['SVM']
    sample_pred = SVM.predict(sample)

    return sample_pred[0]

# Example usage
sample = [6, 148, 72, 35, 0, 33.6, 0.627, 50]

predicted_class = predict_sample(sample)
print(f"Predicted class: {predicted_class}")


Predicted class: 1
